### ISSUE TITLE
Unexpected results I am getting. BUG in the CUDA Simulator.

### ISSUE THREAD

**Issue by caitaozhan** [numba/numba#3265]
## Reporting a bug

-  I am using the version 0.38 of Numba 
-  I have included below a minimal working reproducer.

There is an enexecpted result I am getting when using the <b>CUDA simulator</b>. I write the code on my macbook, use the simulator to debug, then deploy on a linux server with NVIDIA chip. 

```python
import numpy as np
from numba import cuda, float64

local_array_size = 1

def update_size(size):
    global local_array_size
    local_array_size = size

@cuda.jit('void(float64[:])')
def kernal(array):
    x = cuda.grid(1)
    local_array = cuda.local.array(local_array_size, dtype=float64)
    for i in range(local_array_size):
        local_array[i] = i
    array[x] = local_array[local_array_size-1]

d_array = cuda.device_array(32, np.float64)
threadsperblock = 32
blockspergrid = 1
update_size(10)   # <-- this produces the error

kernal[blockspergrid, threadsperblock](d_array)

print(d_array.copy_to_host())

```

On my macbook (no NVIDIA GPU) with NUMBA_ENABLE_CUDASIM=1. The result is the following:
```
[ 9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.]
```

On a linux server with NVIDIA GPU and with NUMBA_ENABLE_CUDASIM=1. The result is the following:
```
[ 9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.  9.]
```

On a linux server with NVIDIA GPU and not simulating. THe result is the following:
```
[ 0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0. ]
```

Therefore, the simulator has some issues. **The result of the simulator does not match the result running on a real GPU.**

I am doing this because I want the size of cuda.local.array to be different from time to time. This seems impossible, because the documentation says that the size of cuda local array must be a simple constant expression. Is there a way to workaround?

**Comment by seibert** [comment]
This is not so much a CUDA simulator bug as a general discrepancy between Numba and Python:

https://numba.pydata.org/numba-doc/dev/user/faq.html#numba-doesn-t-seem-to-care-when-i-modify-a-global-variable

Numba freezes global values at compile time, and your example triggers compilation before update_size() is called by giving an explicit signature (`void(float64[:])`) on the `@cuda.jit` decorator.  This causes the value of `local_array_size` to be captured in the compiled CUDA function as `1`.  The CUDA simulator is intended to allow you to check program logic using `print` and `pdb`, so the code is not compiled at all.  We have plans (thanks to improvements in Python 3.6) for how to get the Numba behavior to match Python, but this will take a little while to get to on the roadmap.

Until then, you should avoid using global variables with Numba (CPU or GPU) unless they are constants.

**Comment by caitaozhan** [comment]
Thanks.

**Comment by stuartarchibald** [comment]
Closing as resolved and #3267 covers the fundamental issue.
